src/bin/pgcopydb/ld_transform.c

```diff
@@ -2609,3 +2669,302 @@ LogicalMessageValueEq(LogicalMessageValue *a, LogicalMessageValue *b)
 	/* makes compiler happy */
 	return false;
 }
+
+
+/*
+ * Identifiers such as schema, table, column comes from various
+ * sources(e.g. wal2json, test_decoding and source catalog) and some of them
+ * already escapes identifiers and few don't.
+ * We need to check if the identifier is already quoted or not before
+ * escaping it.
+ * Whatever we are here is not a fool proof escaping mechanism, but a best
+ * effort to make sure that the identifiers are normalized by quoting them
+ * if it is not already quoted.
+ *
+ * Here is an example:
+ * foo -> "foo"
+ * "foo" -> "foo"
+ * foo"bar -> "foo"bar"
+ * "foo -> ""foo"
+ *
+ * The goal of this normalization is to make sure that the identifiers are
+ * comparable in the context of Hash Table.
+ */
+#define NORMALIZED_PG_NAMEDATA_COPY(dst, src) \
+	{ \
+		int len = strlen(src); \
+		if (src[0] == '"' && src[len - 1] == '"') \
+		{ \
+			strlcpy(dst, src, PG_NAMEDATALEN); \
+		} \
+		else \
+		{ \
+			sformat(dst, PG_NAMEDATALEN, "\"%s\"", src); \
+		} \
+	}
+
+/*
+ * fetchGeneratedColumnsForTable fetches the generated columns set for the given
+ * table "nspname.relname".
+ *
+ * Returns true if the table has generated columns and "columns" is set to the
+ * generated columns set, false otherwise.
+ *
+ * NOTE: There is no error condition, if the cache is NULL, it means that we
+ * don't have any generated columns in the catalog.
+ */
+static bool
+fetchGeneratedColumnsForTable(GeneratedColumnsCache *cache,
+							  const char *nspname,
+							  const char *relname,
+							  GeneratedColumnSet **columns)
```

Comment:
Given the style in the rest of the project, I'm not super happy with this implementation. It looks to me like that caller could have a look at the columns parameter and NULL would indicate that there is no generated columns. We could then use the return value to indicate success, as usual.

Also I think a better name would start with “lookup” because we don't fetch anything to anywhere, we just implement a hash-table lookup, which is in line with the output value being NULL when we don't find anything.